compiler/rustc_expand/src/mbe/macro_rules.rs

```diff
@@ -426,13 +435,33 @@ pub fn compile_declarative_macro(
     // Convert it into `MatcherLoc` form.
     let argument_gram = mbe::macro_parser::compute_locs(&argument_gram);
 
-    let parser = Parser::new(&sess.parse_sess, body, true, rustc_parse::MACRO_ARGUMENTS);
+    let create_parser =
+        |body| Parser::new(&sess.parse_sess, body, true, rustc_parse::MACRO_ARGUMENTS);
+
+    let parser = create_parser(body);
     let mut tt_parser =
         TtParser::new(Ident::with_dummy_span(if macro_rules { kw::MacroRules } else { kw::Macro }));
     let argument_map =
         match tt_parser.parse_tt(&mut Cow::Owned(parser), &argument_gram, &mut NoopTracker) {
             Success(m) => m,
-            Failure(token, _, msg) => {
+            Failure(()) => {
+                let ast::ItemKind::MacroDef(macro_def) = &def.kind else {
```

Comment:
Isn't this a dupe of LL 394-397? What's different between these tokens and those already stored in the `body` local?

Reply:
These here haven't been moved out yet. We pass `Cow::Owned` above which might be a pointless microoptimization but maybe not

Reply:
Hm, alright. I wonder if we can avoid cloning those tokens on 395 and instead do the clone in the body of `create_parser`, just so we don't need to re-match here and to make the code simpler.

Regardless, a comment explaining why we need to re-clone the macro def's tokens would be useful.

r=me with the comment, with or without a simplification of cloning the tokens.